{"metadata":{"kernelspec":{"display_name":"Python 3 (ipykernel)","language":"python","name":"python3"},"language_info":{"name":"python","version":"3.10.16","mimetype":"text/x-python","codemirror_mode":{"name":"ipython","version":3},"pygments_lexer":"ipython3","nbconvert_exporter":"python","file_extension":".py"},"panel-cell-order":["b2b4d5c7-826a-4848-9a08-3dfcf8778d4d"]},"nbformat_minor":5,"nbformat":4,"cells":[{"id":"b2b4d5c7-826a-4848-9a08-3dfcf8778d4d","cell_type":"markdown","source":"# Working directly with files and directories via CLI\n\nI hope everyone is feeling more comfortable with coding and using the command-line interface! Now, let's explore how we can interact directly with plain text files, such as .txt, .tsv, and .csv, using the command line. This skill is especially valuable when working with genomic data, where you often need to count the number of FASTA sequences, inspect file contents, or even make quick edits. The command line provides powerful tools to accomplish all of these tasks efficiently—let’s dive in! \n\n**I want to note that many of these commands will not work on files that are not plain text. Some examples of common file types that are not plain-text files would be “.docx”, “.pdf”, or “.xlsx”. This is because those file formats contain special types of compression and formatting information that are only interpretable by programs specifically designed to work with them.**","metadata":{"panel-layout":{"height":174.140625,"visible":true,"width":100}}},{"id":"f1fb372e-f596-4c26-820f-03763f39aea7","cell_type":"markdown","source":"### Viewing file contents\nNow that we are experts in navigating around our directories, let's make our way to the **unix_intro** folder. \n\nSay I want to see the entire contents of the ***example.txt*** file in our working directory? To do this, I would run the **```cat```** command, which will print and display the entire file content. Let's try it below:","metadata":{}},{"id":"44ed453a-b64a-4c03-843b-9f81da8056a2","cell_type":"code","source":"cat example.txt","metadata":{"trusted":true},"outputs":[{"name":"stdout","output_type":"stream","text":"This is line 1\nThis is a pretend data file\nThis is line 2\nThis is line 3\nThis is line 4\nThis is line 5\nThis is line 6\nThis is line 7\nThere may be fasta sequences here\nThis is line 8\nThis is line 9\nThis is line 10\nThis is line 11\nThis is line 12\nThis is line 13\nThis is line 14\nThis is line 15\nThis is line 16\nThis is line 17\nThis is line 18\nThese could be fasta sequences\nThis is line 19\nThis is line 20\nThis is line 21\nThis is line 22\nThis is line 23\nThis is line 24\nThis is line 25\nThis is line 26\nThis is line 27\nThis is line 28\nThis is line 29\nThis is line 30\nThis is line 31\nThis is line 32\nThis is line 33\nThis is line 34\nThis is line 35\nThis is line 36\nThis is line 37\nThis is line 38\nThis is line 39\nThis is line 40\nThis is line 41\nThis is line 42\nThis is line 43\nThis is line 44\nThis is line 45\nI have data in here\nThis is line 46\nThis is line 47\nThis is line 48\nThis is line 49\nThis is line 50\nThis is line 51\nThis is line 52\nThis is line 53\nThis is line 54\nThis is line 55\nThis is line 56\nThis is line 57\nThis is line 58\nThis is line 59\nThis is line 60\nThis is line 61\nThis is my data\nThis is line 62\nThis is line 63\nThis is line 64\nThis is line 65\nThis is line 66\nThis is line 67\nThis is line 68\nThis is line 69\nThis is line 70\nThis is line 71\nThis is line 72\nThis is line 73\nThis is line 74\nThis is line 75\nThis is line 76\nThis is line 77\nThis is line 78\nThis is line 79\nThis is line 80\nThis is line 81\nThis is line 82\nThis is line 83\nThis is line 84\nThis is line 85\nThis is line 86\nThis is line 87\nThis is line 88\nThis is line 89\nThis is line 90\nThis is line 91\nThis is line 92\nThis is line 93\nI have data in here\nThis is line 94\nThis is line 95\nThis is line 96\nI could have fasta sequences in here\nThis is line 97\nThis is line 98\nThis is line 99\nThis is line 100\n"}],"execution_count":7},{"id":"025a894b-60fb-4c4a-8aa1-5882c0a17066","cell_type":"markdown","source":"In using the ```cat``` command, you will see the entire contents of the text file. Let's number the file lines using the ```-n``` flag.","metadata":{}},{"id":"4e961355-68c4-4086-956f-4563d3a92494","cell_type":"code","source":"cat -n example.txt","metadata":{"trusted":true},"outputs":[{"name":"stdout","output_type":"stream","text":"     1\tThis is line 1\n     2\tThis is a pretend data file\n     3\tThis is line 2\n     4\tThis is line 3\n     5\tThis is line 4\n     6\tThis is line 5\n     7\tThis is line 6\n     8\tThis is line 7\n     9\tThere may be fasta sequences here\n    10\tThis is line 8\n    11\tThis is line 9\n    12\tThis is line 10\n    13\tThis is line 11\n    14\tThis is line 12\n    15\tThis is line 13\n    16\tThis is line 14\n    17\tThis is line 15\n    18\tThis is line 16\n    19\tThis is line 17\n    20\tThis is line 18\n    21\tThese could be fasta sequences\n    22\tThis is line 19\n    23\tThis is line 20\n    24\tThis is line 21\n    25\tThis is line 22\n    26\tThis is line 23\n    27\tThis is line 24\n    28\tThis is line 25\n    29\tThis is line 26\n    30\tThis is line 27\n    31\tThis is line 28\n    32\tThis is line 29\n    33\tThis is line 30\n    34\tThis is line 31\n    35\tThis is line 32\n    36\tThis is line 33\n    37\tThis is line 34\n    38\tThis is line 35\n    39\tThis is line 36\n    40\tThis is line 37\n    41\tThis is line 38\n    42\tThis is line 39\n    43\tThis is line 40\n    44\tThis is line 41\n    45\tThis is line 42\n    46\tThis is line 43\n    47\tThis is line 44\n    48\tThis is line 45\n    49\tI have data in here\n    50\tThis is line 46\n    51\tThis is line 47\n    52\tThis is line 48\n    53\tThis is line 49\n    54\tThis is line 50\n    55\tThis is line 51\n    56\tThis is line 52\n    57\tThis is line 53\n    58\tThis is line 54\n    59\tThis is line 55\n    60\tThis is line 56\n    61\tThis is line 57\n    62\tThis is line 58\n    63\tThis is line 59\n    64\tThis is line 60\n    65\tThis is line 61\n    66\tThis is my data\n    67\tThis is line 62\n    68\tThis is line 63\n    69\tThis is line 64\n    70\tThis is line 65\n    71\tThis is line 66\n    72\tThis is line 67\n    73\tThis is line 68\n    74\tThis is line 69\n    75\tThis is line 70\n    76\tThis is line 71\n    77\tThis is line 72\n    78\tThis is line 73\n    79\tThis is line 74\n    80\tThis is line 75\n    81\tThis is line 76\n    82\tThis is line 77\n    83\tThis is line 78\n    84\tThis is line 79\n    85\tThis is line 80\n    86\tThis is line 81\n    87\tThis is line 82\n    88\tThis is line 83\n    89\tThis is line 84\n    90\tThis is line 85\n    91\tThis is line 86\n    92\tThis is line 87\n    93\tThis is line 88\n    94\tThis is line 89\n    95\tThis is line 90\n    96\tThis is line 91\n    97\tThis is line 92\n    98\tThis is line 93\n    99\tI have data in here\n   100\tThis is line 94\n   101\tThis is line 95\n   102\tThis is line 96\n   103\tI could have fasta sequences in here\n   104\tThis is line 97\n   105\tThis is line 98\n   106\tThis is line 99\n   107\tThis is line 100\n"}],"execution_count":5},{"id":"98799f74-ffa8-406b-95bf-92f6706fe68b","cell_type":"markdown","source":"Now we have numbered all of the lines, so we know exactly how many lines/rows we have in our file. \n\nWhat if we just wanted to see the first few lines of our file? Then, we would use the **```head```** command. Let's try it out:","metadata":{}},{"id":"bc4727f8-1ba5-427f-9561-052c495f8d37","cell_type":"code","source":"head example.txt","metadata":{"trusted":true},"outputs":[{"traceback":["\u001b[0;36m  Cell \u001b[0;32mIn[15], line 1\u001b[0;36m\u001b[0m\n\u001b[0;31m    head example.txt\u001b[0m\n\u001b[0m         ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m invalid syntax\n"],"ename":"SyntaxError","evalue":"invalid syntax (548798623.py, line 1)","output_type":"error"}],"execution_count":15},{"id":"b28ff89f-5c56-451d-b689-5742b21904e4","cell_type":"markdown","source":"What about if we wanted to see the last few lines? We would use the **```tail```** command.","metadata":{}},{"id":"4fd44392-f9de-477d-98d2-ba60abbb0e72","cell_type":"code","source":"tail example.txt","metadata":{"trusted":true},"outputs":[{"traceback":["\u001b[0;36m  Cell \u001b[0;32mIn[16], line 1\u001b[0;36m\u001b[0m\n\u001b[0;31m    tail example.txt\u001b[0m\n\u001b[0m         ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m invalid syntax\n"],"ename":"SyntaxError","evalue":"invalid syntax (834086442.py, line 1)","output_type":"error"}],"execution_count":16},{"id":"ef3390e4-edd6-4a5a-a84b-f8bd8ba94146","cell_type":"markdown","source":"By default, ```head``` and ```tail``` will show the first and last 10 lines of a file, respectively. If you want to see a specific number of lines, we can specify using the ```-n``` flag. \nLet's say I only want to see the first 5 lines of the **example.txt** file. We would run:","metadata":{}},{"id":"5d8a42ab-c18e-4686-bbad-b6c166ff2470","cell_type":"code","source":"head -n 5 example.txt","metadata":{"trusted":true},"outputs":[{"traceback":["\u001b[0;36m  Cell \u001b[0;32mIn[17], line 1\u001b[0;36m\u001b[0m\n\u001b[0;31m    head -n 5 example.txt\u001b[0m\n\u001b[0m            ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m invalid syntax\n"],"ename":"SyntaxError","evalue":"invalid syntax (1130201577.py, line 1)","output_type":"error"}],"execution_count":17},{"id":"aab4d286-7b56-4f61-8d13-6fc37272b2e0","cell_type":"markdown","source":"And what about the last 5 lines?","metadata":{}},{"id":"95ae80f1-935e-42fc-844e-69098914a5a5","cell_type":"code","source":"tail -n 5 example.txt","metadata":{"trusted":true},"outputs":[{"traceback":["\u001b[0;36m  Cell \u001b[0;32mIn[18], line 1\u001b[0;36m\u001b[0m\n\u001b[0;31m    tail -n 5 example.txt\u001b[0m\n\u001b[0m            ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m invalid syntax\n"],"ename":"SyntaxError","evalue":"invalid syntax (4082300057.py, line 1)","output_type":"error"}],"execution_count":18},{"id":"307b99e7-4524-4b99-810d-e01173300e71","cell_type":"markdown","source":"**Great work everyone! These are some of the most basic commands for looking into your files, and some of the commands we use most often in bioinformatics to inspect our data.**\n\nLet's say we have a very large file we want to inspect. If we tries to use the ```cat``` command, it may take up a lot of space in our CLI, or we may crash our environment trying to load it in. In this case, we may use something like the **```less```** or **```more```** commands to inspect the files/\n\n##### ```less``` command\nThe less command allows you to view a file page by page. Let's try it out on the **example.txt** file first (if the notebook spits out the entire file, please move to a new terminal session to visualize). ","metadata":{}},{"id":"e507d3b2-824c-4d0f-a683-596959d00b07","cell_type":"code","source":"less example.txt","metadata":{"trusted":true},"outputs":[{"output_type":"display_data","data":{"text/plain":"This is line 1\nThis is a pretend data file\nThis is line 2\nThis is line 3\nThis is line 4\nThis is line 5\nThis is line 6\nThis is line 7\nThere may be fasta sequences here\nThis is line 8\nThis is line 9\nThis is line 10\nThis is line 11\nThis is line 12\nThis is line 13\nThis is line 14\nThis is line 15\nThis is line 16\nThis is line 17\nThis is line 18\nThese could be fasta sequences\nThis is line 19\nThis is line 20\nThis is line 21\nThis is line 22\nThis is line 23\nThis is line 24\nThis is line 25\nThis is line 26\nThis is line 27\nThis is line 28\nThis is line 29\nThis is line 30\nThis is line 31\nThis is line 32\nThis is line 33\nThis is line 34\nThis is line 35\nThis is line 36\nThis is line 37\nThis is line 38\nThis is line 39\nThis is line 40\nThis is line 41\nThis is line 42\nThis is line 43\nThis is line 44\nThis is line 45\nI have data in here\nThis is line 46\nThis is line 47\nThis is line 48\nThis is line 49\nThis is line 50\nThis is line 51\nThis is line 52\nThis is line 53\nThis is line 54\nThis is line 55\nThis is line 56\nThis is line 57\nThis is line 58\nThis is line 59\nThis is line 60\nThis is line 61\nThis is my data\nThis is line 62\nThis is line 63\nThis is line 64\nThis is line 65\nThis is line 66\nThis is line 67\nThis is line 68\nThis is line 69\nThis is line 70\nThis is line 71\nThis is line 72\nThis is line 73\nThis is line 74\nThis is line 75\nThis is line 76\nThis is line 77\nThis is line 78\nThis is line 79\nThis is line 80\nThis is line 81\nThis is line 82\nThis is line 83\nThis is line 84\nThis is line 85\nThis is line 86\nThis is line 87\nThis is line 88\nThis is line 89\nThis is line 90\nThis is line 91\nThis is line 92\nThis is line 93\nI have data in here\nThis is line 94\nThis is line 95\nThis is line 96\nI could have fasta sequences in here\nThis is line 97\nThis is line 98\nThis is line 99\nThis is line 100\n"},"metadata":{}}],"execution_count":20},{"id":"e819c770-397c-4261-bd4c-994131cb7d30","cell_type":"markdown","source":"What you should be seeing are the first ~50ish or so lines of our **example.txt** file. At the bottom of the page, you will see that you have a blinking black square. Hit the down arrow on your keyboard one time. Now hit the up arrow. \n\nAs you can see, we can navigate around our files this way. We can also scroll on our mouse to move among the data in our file. This is a great way to look at data in our files, without having to:\n* Open it in GUI (when we may not even have an application on our computer that can read the file)\n* Load in all of the data using ```cat```\n\nAfter using the `less` command, you might notice that your terminal doesn’t immediately return to a new command line prompt. This is because `less` allows you to scroll through the file interactively, and it remains open until you explicitly close it. To exit `less` and return to the command prompt, simply press the **\"q\"** key (for \"quit\"). That’s it—just press **q**!\n\nI want to show you what the ```less``` command looks like on a larger file. To do this, lets navigate to the **~/six_commands** directory and try it out on the **example_gene_annotations.csv** file. ","metadata":{}},{"id":"ca1b7f22-e3d3-41c7-a8f9-a6c146216b1b","cell_type":"code","source":"less example_gene_annotations.csv","metadata":{"trusted":true},"outputs":[{"output_type":"display_data","data":{"text/plain":"gene_ID,genome,KO_ID,KO_annotation\n1,CC9311,K02338,DPO3B; DNA polymerase III subunit beta [EC:2.7.7.7]\n2,CC9311,NA,NA\n3,CC9311,K01952,purL; phosphoribosylformylglycinamidine synthase [EC:6.3.5.3]\n4,CC9311,K00764,purF; amidophosphoribosyltransferase [EC:2.4.2.14]\n5,CC9311,K02469,gyrA; DNA gyrase subunit A [EC:5.99.1.3]\n6,CC9311,NA,NA\n7,CC9311,K18979,queG; epoxyqueuosine reductase [EC:1.17.99.6]\n8,CC9311,NA,NA\n9,CC9311,NA,NA\n10,CC9311,K03625,nusB; N utilization substance protein B\n11,CC9311,K03110,ftsY; fused signal recognition particle receptor\n12,CC9311,K07315,rsbU_P; phosphoserine phosphatase RsbU/P [EC:3.1.3.3]\n13,CC9311,K01755,argH; argininosuccinate lyase [EC:4.3.2.1]\n14,CC9311,NA,NA\n15,CC9311,K05539,dusA; tRNA-dihydrouridine synthase A [EC:1.-.-.-]\n16,CC9311,K07305,msrB; peptide-methionine (R)-S-oxide reductase [EC:1.8.4.12]\n17,CC9311,K07007,uncharacterized protein\n18,CC9311,NA,NA\n19,CC9311,K02653,pilC; type IV pilus assembly protein PilC\n20,CC9311,K02669,pilT; twitching motility protein PilT\n21,GEYO,NA,NA\n22,GEYO,K02637,petD; cytochrome b6-f complex subunit 4\n23,GEYO,K02635,petB; cytochrome b6\n24,GEYO,K03797,E3.4.21.102; carboxyl-terminal processing protease [EC:3.4.21.102]\n25,GEYO,K06885,uncharacterized protein\n26,GEYO,K03610,minC; septum site-determining protein MinC\n27,GEYO,K03609,minD; septum site-determining protein MinD\n28,GEYO,K03608,minE; cell division topological specificity factor\n29,GEYO,NA,NA\n30,GEYO,NA,NA\n31,GEYO,K07566,tsaC; L-threonylcarbamoyladenylate synthase [EC:2.7.7.87]\n32,GEYO,K02493,hemK; release factor glutamine methyltransferase [EC:2.1.1.297]\n33,GEYO,K04096,smf; DNA processing protein\n34,GEYO,K04096,smf; DNA processing protein\n35,GEYO,K07107,ybgC; acyl-CoA thioester hydrolase [EC:3.1.2.-]\n36,GEYO,NA,NA\n37,GEYO,K02714,psbM; photosystem II PsbM protein\n38,GEYO,NA,NA\n39,GEYO,K02704,psbB; photosystem II CP47 chlorophyll apoprotein\n40,GEYO,K02718,psbT; photosystem II PsbT protein\n41,N32,NA,NA\n42,N32,NA,NA\n43,N32,NA,NA\n44,N32,NA,NA\n45,N32,NA,NA\n46,N32,NA,NA\n47,N32,K00721,DPM1; dolichol-phosphate mannosyltransferase [EC:2.4.1.83]\n48,N32,NA,NA\n49,N32,NA,NA\n50,N32,NA,NA\n51,N32,K07336,PKHD-type hydroxylase [EC:1.14.11.-]\n52,N32,K02012,afuA; iron(III) transport system substrate-binding protein\n53,N32,NA,NA\n54,N32,K02639,petF; ferredoxin\n55,N32,NA,NA\n56,N32,NA,NA\n57,N32,K00384,trxB; thioredoxin reductase (NADPH) [EC:1.8.1.9]\n58,N32,K03839,fldA; flavodoxin I\n59,N32,NA,NA\n60,N32,NA,NA\n61,RCC307,K02338,DPO3B; DNA polymerase III subunit beta [EC:2.7.7.7]\n62,RCC307,NA,NA\n63,RCC307,K01952,purL; phosphoribosylformylglycinamidine synthase [EC:6.3.5.3]\n64,RCC307,K00764,purF; amidophosphoribosyltransferase [EC:2.4.2.14]\n65,RCC307,K02469,gyrA; DNA gyrase subunit A [EC:5.99.1.3]\n66,RCC307,NA,NA\n67,RCC307,K18979,queG; epoxyqueuosine reductase [EC:1.17.99.6]\n68,RCC307,NA,NA\n69,RCC307,NA,NA\n70,RCC307,K03625,nusB; N utilization substance protein B\n71,RCC307,K03110,ftsY; fused signal recognition particle receptor\n72,RCC307,K07315,rsbU_P; phosphoserine phosphatase RsbU/P [EC:3.1.3.3]\n73,RCC307,K01755,argH; argininosuccinate lyase [EC:4.3.2.1]\n74,RCC307,NA,NA\n75,RCC307,K05539,dusA; tRNA-dihydrouridine synthase A [EC:1.-.-.-]\n76,RCC307,K07305,msrB; peptide-methionine (R)-S-oxide reductase [EC:1.8.4.12]\n77,RCC307,K07007,uncharacterized protein\n78,RCC307,K02653,pilC; type IV pilus assembly protein PilC\n79,RCC307,K02669,pilT; twitching motility protein PilT\n80,RCC307,K02652,pilB; type IV pilus assembly protein PilB\n81,UW179A,NA,NA\n82,UW179A,K02705,psbC; photosystem II CP43 chlorophyll apoprotein\n83,UW179A,NA,NA\n84,UW179A,NA,NA\n85,UW179A,NA,NA\n86,UW179A,K03768,PPIB; peptidyl-prolyl cis-trans isomerase B (cyclophilin B) [EC:5.2.1.8]\n87,UW179A,K01653,E2.2.1.6S; acetolactate synthase I/III small subunit [EC:2.2.1.6]\n88,UW179A,NA,NA\n89,UW179A,K00555,TRMT1; tRNA (guanine26-N2/guanine27-N2)-dimethyltransferase [EC:2.1.1.215 2.1.1.216]\n90,UW179A,K02643,petM; cytochrome b6-f complex subunit 7\n91,UW179A,NA,NA\n92,UW179A,NA,NA\n93,UW179A,NA,NA\n94,UW179A,K06181,rluE; 23S rRNA pseudouridine2457 synthase [EC:5.4.99.20]\n95,UW179A,NA,NA\n96,UW179A,K02518,infA; translation initiation factor IF-1\n97,UW179A,K00384,trxB; thioredoxin reductase (NADPH) [EC:1.8.1.9]\n98,UW179A,NA,NA\n99,UW179A,NA,NA\n100,UW179A,NA,NA\n"},"metadata":{}}],"execution_count":2},{"id":"81eb7596-567b-459b-859b-04709039309a","cell_type":"markdown","source":"##### ```more``` command\nThe ```more``` is very similar to the ```less``` command, but it really only allows for forward navigation. Feel free to try it out in the command line, but since it is limited compared to ```less```, I tend not to use it. ","metadata":{}},{"id":"429c9062-ca9d-45a3-ad43-aeb8699f43d8","cell_type":"markdown","source":"","metadata":{}}]}