{"metadata":{"kernelspec":{"display_name":"Python 3 (ipykernel)","language":"python","name":"python3"},"language_info":{"name":"python","version":"3.10.16","mimetype":"text/x-python","codemirror_mode":{"name":"ipython","version":3},"pygments_lexer":"ipython3","nbconvert_exporter":"python","file_extension":".py"},"panel-cell-order":["b2b4d5c7-826a-4848-9a08-3dfcf8778d4d"]},"nbformat_minor":5,"nbformat":4,"cells":[{"id":"b2b4d5c7-826a-4848-9a08-3dfcf8778d4d","cell_type":"markdown","source":"# Working directly with files and directories via CLI\n\nI hope everyone is feeling more comfortable with coding and using the command-line interface! Now, let's explore how we can interact directly with plain text files, such as .txt, .tsv, and .csv, using the command line. This skill is especially valuable when working with genomic data, where you often need to count the number of FASTA sequences, inspect file contents, or even make quick edits. The command line provides powerful tools to accomplish all of these tasks efficiently—let’s dive in! \n\n*Remember, if you need help with any of these commands further, you can always use the built-in help menus that we can access by providing ```-h``` or ```--help``` as the only argument.*\n\n**I want to note that many of these commands will not work on files that are not plain text. Some examples of common file types that are not plain-text files would be “.docx”, “.pdf”, or “.xlsx”. This is because those file formats contain special types of compression and formatting information that are only interpretable by programs specifically designed to work with them.**","metadata":{"panel-layout":{"height":174.140625,"visible":true,"width":100}}},{"id":"f1fb372e-f596-4c26-820f-03763f39aea7","cell_type":"markdown","source":"## Viewing file contents\nNow that we are experts in navigating around our directories, let's make our way to the **unix_intro** folder. \n\nSay I want to see the entire contents of the ***example.txt*** file in our working directory? To do this, I would run the **```cat```** command, which will print and display the entire file content. Let's try it below:","metadata":{}},{"id":"44ed453a-b64a-4c03-843b-9f81da8056a2","cell_type":"code","source":"cat example.txt","metadata":{"trusted":false},"outputs":[],"execution_count":null},{"id":"025a894b-60fb-4c4a-8aa1-5882c0a17066","cell_type":"markdown","source":"In using the ```cat``` command, you will see the entire contents of the text file. Let's number the file lines using the ```-n``` flag.","metadata":{}},{"id":"4e961355-68c4-4086-956f-4563d3a92494","cell_type":"code","source":"cat -n example.txt","metadata":{"trusted":false},"outputs":[],"execution_count":null},{"id":"98799f74-ffa8-406b-95bf-92f6706fe68b","cell_type":"markdown","source":"Now we have numbered all of the lines, so we know exactly how many lines/rows we have in our file. \n\nWhat if we just wanted to see the first few lines of our file? Then, we would use the **```head```** command. Let's try it out:","metadata":{}},{"id":"bc4727f8-1ba5-427f-9561-052c495f8d37","cell_type":"code","source":"head example.txt","metadata":{"trusted":false},"outputs":[],"execution_count":null},{"id":"b28ff89f-5c56-451d-b689-5742b21904e4","cell_type":"markdown","source":"What about if we wanted to see the last few lines? We would use the **```tail```** command.","metadata":{}},{"id":"4fd44392-f9de-477d-98d2-ba60abbb0e72","cell_type":"code","source":"tail example.txt","metadata":{"trusted":false},"outputs":[],"execution_count":null},{"id":"ef3390e4-edd6-4a5a-a84b-f8bd8ba94146","cell_type":"markdown","source":"By default, ```head``` and ```tail``` will show the first and last 10 lines of a file, respectively. If you want to see a specific number of lines, we can specify using the ```-n``` flag. \nLet's say I only want to see the first 5 lines of the **example.txt** file. We would run:","metadata":{}},{"id":"5d8a42ab-c18e-4686-bbad-b6c166ff2470","cell_type":"code","source":"head -n 5 example.txt","metadata":{"trusted":false},"outputs":[],"execution_count":null},{"id":"aab4d286-7b56-4f61-8d13-6fc37272b2e0","cell_type":"markdown","source":"And what about the last 5 lines?","metadata":{}},{"id":"95ae80f1-935e-42fc-844e-69098914a5a5","cell_type":"code","source":"tail -n 5 example.txt","metadata":{"trusted":false},"outputs":[],"execution_count":null},{"id":"307b99e7-4524-4b99-810d-e01173300e71","cell_type":"markdown","source":"**Great work everyone! These are some of the most basic commands for looking into your files, and some of the commands we use most often in bioinformatics to inspect our data.**\n\nLet's say we have a very large file we want to inspect. If we tries to use the ```cat``` command, it may take up a lot of space in our CLI, or we may crash our environment trying to load it in. In this case, we may use something like the **```less```** or **```more```** commands to inspect the files.","metadata":{}},{"id":"e7b6016d-6c8c-40b9-9ea5-8c1b61a74742","cell_type":"markdown","source":"### ```less``` command\nThe ```less``` command allows you to view a file page by page. Let's try it out on the **example.txt** file first (if the notebook spits out the entire file, please move to a new terminal session to visualize). ","metadata":{}},{"id":"e507d3b2-824c-4d0f-a683-596959d00b07","cell_type":"code","source":"less example.txt","metadata":{"trusted":false},"outputs":[],"execution_count":null},{"id":"e819c770-397c-4261-bd4c-994131cb7d30","cell_type":"markdown","source":"What you should be seeing are the first ~50ish or so lines of our **example.txt** file. At the bottom of the page, you will see that you have a blinking black square. Hit the down arrow on your keyboard one time. Now hit the up arrow. \n\nAs you can see, we can navigate around our files this way. We can also scroll on our mouse to move among the data in our file. This is a great way to look at data in our files, without having to:\n* Open it in GUI (when we may not even have an application on our computer that can read the file)\n* Load in all of the data using ```cat```\n\nAfter using the `less` command, you might notice that your terminal doesn’t immediately return to a new command line prompt. This is because `less` allows you to scroll through the file interactively, and it remains open until you explicitly close it. To exit `less` and return to the command prompt, simply press the **\"q\"** key (for \"quit\"). That’s it—just press **q**!\n\nI want to show you what the ```less``` command looks like on a larger file. To do this, lets navigate to the **~/six_commands** directory and try it out on the **genes_and_seqs.tsv** file. ","metadata":{}},{"id":"ca1b7f22-e3d3-41c7-a8f9-a6c146216b1b","cell_type":"code","source":"less genes_and_seqs.tsv","metadata":{"trusted":false},"outputs":[],"execution_count":null},{"id":"d8ca2f85-788b-4626-ada1-e87080ea3eff","cell_type":"markdown","source":"***Here, we can see that the ```less``` command outputed information about the number of genes (gene_ID) we have in our dataset, the length of the sequence (AA_length), and the protein sequence that corresponds to that gene (seq).***","metadata":{}},{"id":"81eb7596-567b-459b-859b-04709039309a","cell_type":"markdown","source":"##### ```more``` command\nThe ```more``` is very similar to the ```less``` command, but it really only allows for forward navigation. Feel free to try it out in the command line, but since it is limited compared to ```less```, I tend not to use it. ","metadata":{}},{"id":"fc7bd0e9-73eb-4b28-9bc5-9c273fa78a22","cell_type":"markdown","source":"### Word count- ```wc```\n\n##### Basic Function\n```wc``` counts lines, words, and characters in text files. It's basic syntax is: ```wc [options] [file(s)]```. Let's try it on the **example.txt** file.","metadata":{}},{"id":"13e37f20-43ab-41f6-b540-abe0835598da","cell_type":"code","source":"wc example.txt","metadata":{"trusted":false},"outputs":[{"traceback":["\u001b[0;36m  Cell \u001b[0;32mIn[2], line 1\u001b[0;36m\u001b[0m\n\u001b[0;31m    wc example.txt\u001b[0m\n\u001b[0m       ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m invalid syntax\n"],"ename":"SyntaxError","evalue":"invalid syntax (353116505.py, line 1)","output_type":"error"}],"execution_count":2},{"id":"539a84dd-ea45-4daa-b1c2-cc9ffe0acb74","cell_type":"markdown","source":"The ```wc``` command outputs three different numbers:\n1. Number of lines\n2. Number of words\n3. Number of bytes (characters)\n\nSo in the **example.txt** file, we have 107 lines, 438 words, and the file is 1778 bytes. \n\n##### Common Options\n* ```-l```: Count only lines\n* ```-w```: Count only words\n* ```-c```: Count bytes\n* ```-m```: Count characters\n\nThis command has many applications in bioinformatics research. ","metadata":{}},{"id":"429c9062-ca9d-45a3-ad43-aeb8699f43d8","cell_type":"markdown","source":"### Searching for patterns- ```grep```","metadata":{}},{"id":"619efa60-e7d4-4d35-ad06-274170d548f3","cell_type":"markdown","source":"When searching for specific words or phrases in a Word document or PDF, my go-to method is **CTRL+F**, which quickly locates the information I need without scrolling through unnecessary content. **CTRL+F** is the graphical interface (GUI) tool for pattern searching in documents. In the command-line interface (CLI), we use the **`grep`** command to achieve the same functionality when working with plain text files. This tool is very versatile, so let's take our time learning it. ","metadata":{}},{"id":"634e740b-a362-4bea-97bb-c6a4b87ea460","cell_type":"markdown","source":"##### Basic Function\n```grep``` stands for \"globally search a regular expression and print\". It searches for specific patterns in files and displays lines containing those patterns.\n\n##### Syntax\nThe basic syntax of grep is:\n```grep [options] pattern [file(s)]```\n\n##### Common Options\n* ```-i```: Ignore case distinctions\n\n* ```-v```: Invert the match (show lines that don't match)\n\n* ```-c```: Count matching lines instead of printing them\n\n* ```-n```: Display line numbers along with matching lines\n\n* ```-w```: Match whole words only\n\n##### Bioinformatics Applications\nIn bioinformatics, we use ```grep``` extensively for multiple different purposes. These include:\n* Searching Sequence Files\n* Extracting FASTA Headers\n* Counting Sequences\n\nLet's try it with our files!","metadata":{}},{"id":"3b5fdb5a-837d-4cce-9dd6-13cebeaf4fd1","cell_type":"markdown","source":"First off, let's say I want to see how many \"blue\" are in my **colors.txt** file. To do this, we would simply run: ","metadata":{}},{"id":"912e33d8-2d01-435c-a69f-f9536ab8fddf","cell_type":"code","source":"grep blue colors.txt","metadata":{"trusted":false},"outputs":[],"execution_count":null},{"id":"c8d79a3e-ccb7-4367-b69b-96d005716d05","cell_type":"markdown","source":"This should print out the lines that contain what you searched for. What if what we are searching for has multiple hits in a single file?","metadata":{}},{"id":"3ffeb615-05de-4359-b088-32f3fc2edb69","cell_type":"code","source":"grep re colors.txt","metadata":{"trusted":false},"outputs":[],"execution_count":null},{"id":"ec041701-7263-44df-bd77-cc5cfa374a11","cell_type":"markdown","source":"You'll see that grep gave us two outputs: **red** and **green**, as they both contain **re**. \n\nLet's try with another file, **example.txt**","metadata":{}},{"id":"4db5d430-1ea1-4d1e-a3ea-3ef511f63569","cell_type":"code","source":"grep data example.txt","metadata":{"trusted":false},"outputs":[],"execution_count":null},{"id":"62464b6d-da10-4bc7-ae5c-2d8486999845","cell_type":"markdown","source":"From this output, we can see that the word **\"data\"** appears four times in our file. Additionally, instead of simply displaying the number of occurrences, the command prints the entire lines where the word appears. For example, if the file contains the sentence **\"This is a pretend data file,\"** that full line will be shown in the results. \n\nWhat if we want to see the line numbers on which the word **\"data\"** occurs? We would use the ```-n``` flag:","metadata":{}},{"id":"a00d65f5-5c93-4674-9d1e-c3ce3b75df95","cell_type":"code","source":"grep -n data example.txt","metadata":{"trusted":false},"outputs":[],"execution_count":null},{"id":"291e009c-4a66-485f-a9d0-917fbac754fa","cell_type":"markdown","source":"Now we have the line #s that the **data** occurs on. ","metadata":{}},{"id":"12eafa98-c5ee-45dc-ade2-2229c641df50","cell_type":"markdown","source":"We if we ***just*** wanted the number of times **data** occurs without seeing the full text in the line? We would use the **```-c```** flag:","metadata":{}},{"id":"564eb3f8-7467-4107-b375-8703cd8f82de","cell_type":"code","source":"grep -c data example.txt","metadata":{"trusted":false},"outputs":[],"execution_count":null},{"id":"fbe57249-8cf2-4357-a0d7-806c25e3831f","cell_type":"markdown","source":"This just prints out the number of **data** occurences in the file.\n\nA great use for ```grep -c``` is counting the number of sequences you may have in a .fasta file. In .fasta file, sequences are characterized by identifying header lines that give you general information about the sequence. A header *always* starts with **'>'**. An example of this looks like:\n\n    >chr1 Jackalope chromosome 1;length=7\n    GATTACA\n\nSo by using the ```grep``` command with the '>' symbol, we can count how many sequences are in a .fasta file. Let's try it with the **sequence.fasta** file. ","metadata":{}},{"id":"77d088ac-ef9b-4c7e-abef-19b936cbdc81","cell_type":"code","source":"grep -c \">\" sequence.fasta","metadata":{"trusted":true},"outputs":[{"traceback":["\u001b[0;36m  Cell \u001b[0;32mIn[1], line 1\u001b[0;36m\u001b[0m\n\u001b[0;31m    grep -c \">\" sequence.fasta\u001b[0m\n\u001b[0m            ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m invalid syntax\n"],"ename":"SyntaxError","evalue":"invalid syntax (1982783783.py, line 1)","output_type":"error"}],"execution_count":1},{"id":"733e91c7-791e-4302-a1b3-95585c1cef43","cell_type":"markdown","source":"This should tell you that there are 2 sequences in the **sequence.fasta** file. Let's take a look to make sure that it's true!","metadata":{}},{"id":"0f67a59f-1176-4144-8b29-ee5d39e57c9a","cell_type":"code","source":"less sequence.fasta","metadata":{"trusted":true},"outputs":[{"output_type":"display_data","data":{"text/plain":">Seq1 [organism=Homo sapiens] [gene=BRCA1] Homo sapiens BRCA1 gene, partial sequence\nATGGATTTATCTGCTCTTCGCGTTGAAGAAGTACAAAATGTCATTAATGCTATGCAGAAAATCTTAGAGTGTCCC\nATCTGTCTGGAGTTGATCAAGGAACCTGTCTCCACAAAGTGTGACCACATATTTTGCAAATTTTGCATGCTGAA\nACTTCTCAACCAGAAGAAAGGGCCTTCACAGTGTCCTTTATGTAAGAATGATATAACCAAAAGGAGCCTACAAG\nAAAGTACGAGATTTAGTCAACTTGTTGAAGAGCTATTGAAAATCATTTGTGCTTTTCAGCTTGACACAGGTTTG\nGAAAAAACAAATACTGAAATTAGATAGGCTGGAATTACAGAAAAAGTAAAAAAAGAACTATTTATGAAGCAGTG\n\n>Seq2 [organism=Mus musculus] [gene=p53] Mus musculus tumor protein p53 (Trp53) mRNA, partial cds\nATGGAGGAGTCACAGTCGGATATCAGCCTCGAGCTCCCTCTGAGCCAGGAGACATTTTCAGGCTTATGGAAACT\nACTTCCTCCAGAAGATATCCTGCCATCACCTCACTGCATGGACGATCTGTTGCTGCCCCAGGATGTTGAGGAGT\nTTTTTGAAGGCCCAAGTGAAGCCCTCCGAGTGTCAGGAGCTCCTGCAGCACAGGACCCTGTCACCGAGACCCCG\nGACGATATTGAACAATGGTTCACTGAAGACCCAGGTCCAGATGAAGCTCCCAGAATGCCAGAGGCTGCTCCGAT\nGGTGATGGCCTGGCTCCTCCCCAGCATCTTATCCGGGTGGAAGGAAATTTGTATCCCGAGTATCTGGAAGACAG\n"},"metadata":{}}],"execution_count":3},{"id":"b34982e9-4cf1-4e6f-8399-2efb6861523f","cell_type":"markdown","source":"Great! We see that we do indeed have 2 sequences. \n\nOkay, let's apply this to some more sequence data. Go ahead and move back into your ~/six_commands working directory.\n\nNow, I want you guys to go ahead and look at the header for the **gene_annotations.tsv** file. Remember, we use the **```head```** command to look at the first lines of a file:","metadata":{}},{"id":"a6e6b9dc-529b-4be9-b688-305ecf8bacc8","cell_type":"code","source":"head -n 1 gene_annotations.tsv","metadata":{"trusted":false},"outputs":[],"execution_count":null},{"id":"483c731a-d942-40b9-b755-b69a4a9fe07d","cell_type":"markdown","source":"Does everyone see that we have 4 columns in this file? We have: *gene_ID*, *genome*, *KO_ID*, and *KO_annotation*\n(KO stands for KEGG Orthology; A KO ID (KEGG Orthology ID) is an identifier used in the KEGG (Kyoto Encyclopedia of Genes and Genomes) database to classify genes and proteins based on their functions. We can go over this in more detail if need be).\n\nLet's say that we are hypothetically interested in genes involved with sulfide oxidation, which is thr process of turning H2S (sulfide) into elemental sulfur. Using the KEGG pathway database (we will go over this another time), I see that there is one gene involved in that process: **K17229**.\n\nLet's see if we have this gene in our table!","metadata":{}},{"id":"658a5282-8922-43ae-a17f-8fffc9fc78f2","cell_type":"code","source":"grep K17229 gene_annotations.tsv","metadata":{"trusted":false},"outputs":[],"execution_count":null},{"id":"8f32f1b0-9a87-4ae1-8d42-d100d700671f","cell_type":"markdown","source":"Since this command gave us nothing back, we can assume that this gene is not in our annotated file and was therefore not characterized in our hypothetical study. What if we were instead interested in some of the photosystem II genes that encode important proteins for photosynthesis? I can see from KEGG that I am interested in these KEGG IDs:\n\n* K02703\n* K02706\n* K02705\n* K02704\n* K02707\n* K02708\n\n **With the ```-e``` flag, we can search for multiple variables at a time in a file with a single command.** So let's search for them and see what we got! ","metadata":{}},{"id":"db16f53d-432b-4735-833a-19e86ffe8b79","cell_type":"code","source":"grep -e K02703 -e K02706 -e K02705 -e K02704 -e K02707 -e K02708 gene_annotations.tsv","metadata":{"trusted":false},"outputs":[],"execution_count":null},{"id":"d6a959a3-6219-4764-979e-88ff6394623e","cell_type":"markdown","source":"From this, we can see that we have 2 photosystem II genes in our hypothetical study:\n\n* 39      GEYO    K02704  psbB; photosystem II CP47 chlorophyll apoprotein\n* 82      UW179A  K02705  psbC; photosystem II CP43 chlorophyll apoprotein\n\n***Great work everyone! Getting the hang of these commands is tough. Remember that we are not here to memorize every detail about how to use the CLI. We are just getting a hang of the tools and commands we can use in bioinformatics.***","metadata":{}},{"id":"a6bd31f2-aa7f-4b4d-8b02-c877484dd322","cell_type":"code","source":"","metadata":{"trusted":false},"outputs":[],"execution_count":null}]}